[GPII-3367, GPII-3368]: Rotate secrets, secret keys, Terraform state key

natarajaya opened the pull request:
Added universal `:rotate_secret` task with `:rotate_tfstate_key` wrapper.
Added `:rotate_secrets_key` task and wrapper.

To rotate a secret:
```
rake plain_sh["rake rotate_secret['default'\,'secret_couchdb_admin_username']"]
```

To rotate secrets key:
```
rake plain_sh["rake rotate_secrets_key['default']"]
```

Also, as discussed before, I reorganized xk rakefiles, please share your opinions on new structure!

natarajaya commented:
Added `:rotate_secrets_key` wrapper.

natarajaya commented on shared/rakefiles/xk_util.rake:
```diff
@@ -0,0 +1,36 @@
+# Old secret value will be set to TF_VAR_secret_name_rotated until rotation is finished.
+#
+# Arbitrary command to execute after rotation can be set with :cmd argument.
+task :rotate_secrets, [:encryption_key, :secret, :cmd] => [:configure_serviceaccount] do |taskname, args|
```
I am going to split this in couple simpler functions soon.

natarajaya commented on shared/rakefiles/entrypoint.rake:
```diff
@@ -177,8 +178,23 @@ task :destroy_tfstate, [:prefix] => [:set_vars, :check_destroy_allowed] do |task
+  sh "#{@exekube_cmd} rake rotate_secret['key_tfstate_encryption_key','default','sh -c \"gsutil \
+       -o GSUtil:decryption_key1=$TF_VAR_key_tfstate_encryption_key_rotated \
+       -o GSUtil:encryption_key=$TF_VAR_key_tfstate_encryption_key \
+       rewrite -k -r gs://#{ENV["TF_VAR_project_id"]}-tfstate/#{@env}/#{prefix}\"',skip_secret_mgmt,preserve_stderr]"
```
Because rotating TF state key is just a special case of rotating a secret, and I decided to make `:rotate_secret` task universal. It rotates a secret by default, but when `:cmd` argument is present, it also executes command from that argument to handle cases when we need some additional action at the moment when we have both old and new secrets set as ENV vars.

natarajaya commented on shared/rakefiles/xk_util.rake:
```diff
@@ -0,0 +1,36 @@
+    ENV["TF_VAR_#{args[:secret]}"] = ""
+  end
+
+  Secrets.rotate_key(args[:encryption_key])
```
I agree that KMS key rotation is a potentially dangerous operation, mostly because there is no way to wrap it up into a single "transaction" that can be reverted automatically in case of `:rotate_secrets_key` task gets terminated / loses internet connectivity in the middle of the rotation.

However, there is also no way to irreversibly lose any sensitive data if situation from previous point happens. Disabled KMS key version can be re-enabled again in case we stuck with the secrets file encrypted with it. And secrets GS bucket is versioned in case we need to recover previous version of secrets file.

I experimented with all sorts of interruptions to `:rotate_secrets_key`, `:rotate_secret`  tasks and I was able to recover from any of failures with minimum effort.

natarajaya commented:
@stepanstipl You are right about automatic KMS keys rotation. This is not an option, at least until GPII-3408 is implemented.

mrtyler commented on shared/rakefiles/secrets.rb:
```diff
@@ -119,12 +123,28 @@ def self.push_secrets(secrets, encryption_key)
+      encryption_key_version = JSON.parse(encryption_key_version)
+      encryption_key_version = encryption_key_version['primary']['name'].match(/\/([0-9]+)$/)[1]
+    rescue
+      puts "ERROR: Unable to get primary encryption key version for key '#{encryption_key}', terminating!"
```
If it's safe to do so, it might be helpful to emit the JSON blob as part of this failure to assist debugging.

natarajaya replied:
Done.

mrtyler commented on shared/rakefiles/secrets.rb:
```diff
@@ -119,12 +123,28 @@ def self.push_secrets(secrets, encryption_key)
+
+    begin
+      encryption_key_version = JSON.parse(encryption_key_version)
+      encryption_key_version = encryption_key_version['primary']['name'].match(/\/([0-9]+)$/)[1]
```
I find this hard to understand. Some things that might help me:
* An example of what `primary.name` looks like in the returned JSON blob (`"name": "projects/gpii-gcp-dev-tyler/locations/global/keyRings/keyring/cryptoKeys/default/cryptoKeyVersions/1"`).
* Moving the `['name'].match(/\/([0-9]+)$/)[1]` part (which is repeated three times in this file) to a function `get_latest_version_from_key_name`.

(In general I prefer changing the delimiter over escaping, but here I'm undecided on whether I think it's harder to read `%r{}` or a single `\/`.)

natarajaya replied:
Good suggestion!

mrtyler commented on shared/rakefiles/secrets.rb:
```diff
@@ -207,6 +227,65 @@ def self.fetch_secrets(encryption_key)
+      new_version = JSON.parse(new_version)
+      new_version_id = new_version["name"].match(/\/([0-9]+)$/)[1]
+    rescue
+      puts "ERROR: Unable to create new version for key '#{encryption_key}', terminating!"
```
If it's safe to do so, it might be helpful to emit the JSON blob as part of this failure to assist debugging.

natarajaya replied:
Done.

mrtyler commented on shared/rakefiles/secrets.rb:
```diff
@@ -207,6 +227,65 @@ def self.fetch_secrets(encryption_key)
+    begin
+      key_versions = JSON.parse(key_versions)
+    rescue
+      puts "ERROR: Unable to retrieve versions for key '#{encryption_key}', terminating!"
```
If it's safe to do so, it might be helpful to emit the JSON blob as part of this failure to assist debugging.

natarajaya replied:
Done.

mrtyler commented on shared/rakefiles/secrets.rb:
```diff
@@ -207,6 +227,65 @@ def self.fetch_secrets(encryption_key)
+        version_disabled = JSON.parse(version_disabled)
+        raise unless version_disabled['state'] == "DISABLED"
+      rescue
+        puts "ERROR: Unable to disable version #{version_id} for key '#{encryption_key}', terminating!"
```
If it's safe to do so, it might be helpful to emit the JSON blob as part of this failure to assist debugging.

natarajaya replied:
Done.

natarajaya commented on shared/rakefiles/entrypoint.rake:
```diff
@@ -177,8 +178,34 @@ task :destroy_tfstate, [:prefix] => [:set_vars, :check_destroy_allowed] do |task
+       -o GSUtil:decryption_key1=$TF_VAR_key_tfstate_encryption_key_rotated \
+       -o GSUtil:encryption_key=$TF_VAR_key_tfstate_encryption_key \
+       rewrite -k -r gs://#{ENV["TF_VAR_project_id"]}-tfstate/#{@env}/#{prefix}\"',skip_secret_mgmt,preserve_stderr]"
+  sh "docker volume rm -f -- #{ENV["TF_VAR_project_id"]}-#{ENV["USER"]}-terragrunt"
```
Alright.

natarajaya commented:
@mrtyler @stepanstipl I addressed Tyler's suggestions and also restructured code a little to improve order of operations for less "dangerous" key rotation.
I have a question – do you guys think we need any locking mechanics (something similar to TF state locking, simple gs://gpii-gcp-dev-natarajaya-default-secrets/secret-mgmt.lock?) for this kind of operations? 
cc @amatas

stepanstipl commented:
Thanks, the order of operation looks better now.

Re. locking - it's a good point. I don't think we need any special locking mechanism for key rotation explicitly, it should be run only on CI anyway (for non-dev envs), but in a long term I would like to see a locking mechanism for the whole `rake` run - currently, we have locking mech. for individual parts, given locking of TF state, but that does not really guarantee that multiple runs won't happen at the same time leaving the cluster in inconsitent state.

amatas commented:
Well, right now we have a "lock" that is the number of process that run in the CI which is 1. So no more than one step with a particular tag at the same time should run.

The PR LGTM.

natarajaya commented:
@stepanstipl @amatas Thanks for your replies!
@mrtyler If you share team opinion about locking and don't have other topics to address in scope of this PR, I think it is ready to be merged.

mrtyler commented on shared/rakefiles/secrets.rb:
```diff
@@ -242,12 +242,18 @@ def self.rotate_key(encryption_key)
+  end
+
+  # This method disables all versions except primary for target encryption_key
+  def self.disable_key_versions(encryption_key, primary_version_id = 0)
```
I think `primary_version_id` should be required, not optional. I don't think it makes sense to call this method without a specific version id.

natarajaya replied:
It makes sense to call this method without specific version id if you want to disable kms key entirely. But yeah, yagni this!

mrtyler commented:
GCP's revision controls on keys and buckets are helpful, but I prefer the improved, safer order of operations.

I agree with Stepan that component-level locks are hardly bulletproof[1]. I agree with Alfredo that for now CI limits (plus manual coordination for e.g. making changes to prd) seem good enough. I would prefer that we not implement our own locking logic if we can avoid it.

A downside of our custom secrets management code is that we're forced to extend it to handle new scenarios like key rotation. We have good reasons for being where we are, and I'm not proposing anything different (other than perhaps moving to Vault one day -- still not something I think we should do today). However, I wanted to note this situation as an example of why I like for us to avoid custom code when possible.

[1] aws/ is set up this way. In my early testing, it was easy to get into a situation where one pipeline was deploying an environment while another pipeline was tearing down that same environment.

mrtyler commented:
One more idea:

What do you think of exercising this code in CI somewhere? We could have a dedicated bucket that we rotate on each dev run. We could rotate dev's tfstate on each dev run. Maybe that's too much churn / file io (that we have to pay for), and instead we could rotate stg and prd on each pipeline run (this would replace time-based rotation, e.g. rotating prd tfstate every 60 days, if that's something we want).

Basically, I'd like to see this new code exercised so that we can depend on it the next time we leak a secret ;).

stepanstipl commented:
> A downside of our custom secrets management code is that we're forced to extend it to handle new scenarios like key rotation. We have good reasons for being where we are, and I'm not proposing anything different (other than perhaps moving to Vault one day -- still not something I think we should do today). However, I wanted to note this situation as an example of why I like for us to avoid custom code when possible.

Should we use Google's KMS Customer Managed Keys model, key rotation is supported out of the box, even in automated fashion (set how often the keys should be rotated and KMS does it for you).

natarajaya commented on shared/rakefiles/secrets.rb:
```diff
@@ -133,13 +133,14 @@ def self.push_secrets(secrets, encryption_key)
-      encryption_key_version = encryption_key_version['primary']['name'].match(/\/([0-9]+)$/)[1]
+      encryption_key_version = get_crypto_key_version(encryption_key_version['primary']['name'])
     rescue
+      puts encryption_key_version
```
You are right, let me improve this.

natarajaya commented on shared/rakefiles/secrets.rb:
```diff
@@ -242,12 +242,18 @@ def self.rotate_key(encryption_key)
+  end
+
+  # This method disables all versions except primary for target encryption_key
+  def self.disable_key_versions(encryption_key, primary_version_id = 0)
```
`disable_non_primary_key_versions` it is.

natarajaya commented:
@mrtyler @stepanstipl While I completely agree with your points about custom code, I would like to add that those "good reasons why we are where we are"  were mostly implementational (lack of KMS key support for GS in Terraform) and security (in-memory handling of secret material) issues. 

Since first is not a problem anymore, current secret management implementation can be made much  simpler by getting rid of client-side encryption / decryption / manual key rotation. Of course we would still need to collect / populate / fetch / push secrets. We have [the ticket](https://issues.gpii.net/browse/GPII-3408) about it. One of the other possible improvements would be to use [google-api-client](https://github.com/googleapis/google-api-ruby-client) gem instead of direct API calls. I added this idea to the same ticket.

> What do you think of exercising this code in CI somewhere?

I would say adding this to stg and prd makes sense.

natarajaya commented:
Updated GitLab config, fixed one small issue with secrets file being generated / uploaded / fetched from GS even when there are no secrets for target encryption key.

mrtyler commented on .gitlab-ci.yml:
```diff
@@ -283,6 +283,9 @@ gcp-stg:
     # Remove all SA keys except current one to prevent hitting 10 keys per SA limit (GPII-3299)
     - rake destroy_sa_keys || true
+    # Rotate Terraform state key and secrets KMS key
+    - rake rotate_tfstate_key # We want CI to stop in case any of these after scripts fail
```
I don't think this belongs in after_script. This doesn't seem like something we want to run regardless of whether the build was successful -- it's not a "cleanup" step.

Furthermore, I think this should happen earlier in the Job, right after `configure_serviceaccount_ci_restore`. That way, if rotating tfstate or secrets breaks something, we'll find out in the build where the failure occurred instead of in the next build.

natarajaya replied:
Right, I am not thinking clearly on Wednesdays!

natarajaya commented on shared/rakefiles/entrypoint.rake:
```diff
@@ -86,22 +86,22 @@ task :deploy => [:set_vars, :apply_infra] do
 end
 
-task :check_destroy_allowed do
+task :check_run_allowed do
```
Not removing, renaming for consistency. AWS guard rail named RAKE_REALLY_RUN_IN_PRD, and we also require it for [test tasks](https://github.com/gpii-ops/gpii-infra/blob/master/.gitlab-ci.yml#L351-L352), so it makes sense to me.

natarajaya replied:
Right, we have two guard rails for prd. Nice!

stepanstipl commented:
@natarajaya what you said sounds good to me

> Of course we would still need to collect / populate / fetch / push secrets

Why? If we generate all secrets only with TF, those will only be in encrypted TF state (where they are already anyway). I think that way we're able to completely switch to KMS Customer Managed keys model and would be able to drop any secrets mgmt code while still supporting features like key rotation. Am I missing something?

natarajaya commented:
> Why? If we generate all secrets only with TF, those will only be in encrypted TF state
@stepanstipl If we don't have control over secrets on ruby level, we will not be able to filter them in Terraform output.

stepanstipl commented:
thanks @natarajaya, I think we should try to think if we can solve the filtering problem (i have some ideas), but I guess this is not near priority after yesterday's backlog refinement. I just would like to have the opportunity to do/discuss this before we start any implementation effort.
